microsetta_public_api/config.py

```diff
@@ -350,17 +350,10 @@ def schema(self):
                         "type": "object",
                         "properties": {
                             self.metadata_kw: metadata_schema,
-                        },
-                        "additionalProperties":
-                            {
-                                "type": "object",
-                                "properties": {
-                                    self.alpha_kw: alpha_group_schema,
-                                    self.beta_kw: beta_group_schema,
-                                    self.taxonomy_kw: taxonomy_group_schema,
-                                    self.pcoa_kw: pcoa_group_schema,
```

Comment:
@wasade Add `self.metadata_kw: metadata_schema` instead and I think that should fix your errors. The way it is currently configured now with `additionalProperties` allows for arbitrarily named datasets by allowing for arbitrary keys where the value matches the schema with the `alpha`, `beta`, etc. properties. The proposed change removes the dataset name level.

Reply:
Wouldn't having `self.metadata_kw: metadata_schema` directly under "datasets" as well as within "additionalProperties" allow for metadata to be both dataset independent and dataset specific? It is plausble there is something fundamental I don't understand here

Reply:
Sure. If we want to disallow dataset independent metadata, then we would also remove the `self.metadata_kw: metadata_schema` under the `properties` tag. But we would still need the `additionalProperties` keywords, which have been deleted here.

One of the tests on `CompatibilitySchema` (this also brings up the point that the `CompatibilitySchema, which is the currently used schema by the api, should also be updated unless we are committing to refactoring the existing config before we can re-deploy) is failing because an unexpected keyword `__metadata__` has been added to an `additionalProperty` of `datasets`. The properties of `datasets` are `16SAmplicon` and `ShotgunMetagenomics`, and they are additional properties because they are not pre-specified. And the schema how it is currently defined (before this change) requires that values of those properties, e.g.,
```
                {
                    '__alpha__': {
                        'faith_pd': 'path/to/shotgun_faith_pd.qza',
                    },
                    '__metadata__': '/path/to/metadata.tsv',
                },
```
Only have properties `__alpha__`, `__beta__`, `__pcoa__`, `__taxonomy__`.

The current proposed change instead says that the properties of `datasets` should be the `__alpha__`, `__beta__`, etc. so we lose the dataset resolution. 


The other test is because the proposed change no longer requires the values of the additional properties to match a schema, so an arbitrary additional property's value can have a misc property and not fail validation.

Reply:
I think it would be great to remove the compatibility schema -- reducing the complexity of the configuration system would be pleasant :) To be honest, I don't fully follow this, but I'll make some more pokes and see what happens

Reply:
I agree, but that is independent of the error here. That just happens to be the test case where the error got caught.

Reply:
To give a more concrete example.

Take the schema:
```
{
  "type": "object",
  "properties": {
    "foo": { "type": "string"},
    "bar": {"type": "int"},
  },
  "additionalProperties": False
}
```

Then by this, `{"foo": "qux", "bar": 725}` is valid and `{"foo": "qux", "bar" 725", "beta": {"date": "Jan7"}}` is invalid because it has an additional property 'beta' that is not allowed because we set `"additionalProperties": False`. If we had set `"additionalProperties": True` (`True` by default) then the second example would be valid.

If we set `additionalProperties` to itself be a schema, then each additional property, such as `beta`, must match this schema. For an alternative schema:
```
{
  "type": "object",
  "properties": {
    "foo": { "type": "string"},
    "bar": {"type": "int"},
  },
  "additionalProperties": {
    "type": "object",
    "properties": {"date": {"type": "string"}},
  }
}
```

`{"foo": "qux", "bar" 725", "beta": {"date": "Jan7"}}` would indeed be valid. As would `{"foo": "qux", "bar" 725", "gamma": {"date": "Sept9"}}` and  `{"foo": "qux", "bar" 725", "beta": {"date": "Jan7"}, "gamma": {"date": "Sept9"}}`. Because the additional properties, ('beta' and 'gamma'), have values (`{"date": "Jan7"}` and `{"date": "Sept9"}`) that match the allowable schema for additional properties.

This is different than the following schema:
```
{
  "type": "object",
  "properties": {
    "foo": { "type": "string"},
    "bar": {"type": "int"},
    "date": {"type": "string"}
  }
}
```

Because this is saying that the top level object should have a property named "date".

We are using the "additionalProperties" to arbitrarily add datasets with a given key name, so we need the schema closer to the second example from this comment, where a dataset name are akin to "beta" and "gamma".

Reply:
Thanks! But, why would alpha/beta/taxonomy be additional properties rather than just properties? It sounds like this is effectively making things required or optional, and we might as well require these entries?

Reply:
We can sort that out later though, important thing right now is just that the metadata are dataset specific. Still hoping to poke at it today. Next window is in roughly 75 miuntes

Reply:
They are not additional properties. They are just properties. The _datasets_

Reply:
For clarity we could write,
```
dataset_schema = {
                                "type": "object",
                                "properties": {
                                    self.metadata_kw: metadata_schema
                                    self.alpha_kw: alpha_group_schema,
                                    self.beta_kw: beta_group_schema,
                                    self.taxonomy_kw: taxonomy_group_schema,
                                    self.pcoa_kw: pcoa_group_schema,
                                },
                                "additionalProperties": False,
                            }
```
And then, so that 'datasets' has arbitrary properties that correspond to a dataset each, we could write something like:
```
schema = {
  "properties": {
    "datasets": {
      "additionallProperties": dataset_schema
    }
  }
}
```
So then all of the properties of dataset, which are named "16S", "Metagenomics" for example, have to point to things that look like datasets.

Reply:
Okay, cool cool, this helps a lot. Thanks!! Sorry for being slow on these bits